metrics/prometheus/metrics.go

```diff
@@ -0,0 +1,37 @@
+package prometheus
+
+import "github.com/prometheus/client_golang/prometheus"
+
+type Metrics struct {
+	Counter      *prometheus.CounterVec
+	ResponseTime *prometheus.HistogramVec
+}
+
+func NewMetrics(version, hash, buildTime string) *Metrics {
+	labels := map[string]string{
+		"version":   version,
+		"hash":      hash,
+		"buildTime": buildTime,
+	}
+	pm := &Metrics{
+		Counter: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name:        "sts_requests_total",
+				Help:        "Secure token service requests served per endpoint",
+				ConstLabels: labels,
+			},
+			[]string{"endpoint"},
+		),
+		ResponseTime: prometheus.NewHistogramVec(
```

Comment:
How do we decide which prometheus statistics to use here? Is `sts_response_time_seconds` standardized somewhere, or chosen arbitrarily?

Reply:
Just as an example, you want to name this `secure_token_service`. I'd disagree and say to use the name of the product. Someone else might want to have here `oauth2_server` or whatever. It's really hard to define something that is so dependent on the context it's being used in.

Reply:
I agree with you. Some configurable value can be added (for example CLUSTER_NAME, default hydra) for naming. In this case metrics name will be CLUSTER_NAME_response_time_seconds for example

Reply:
But here we have the same issue, right? Who decides:

a. What metrics to include
b. What the key ("name") is for each metric

In your specific example, I'd say (just to prove the point) that I'd like to have `response_time_ms` because if an endpoint requires more than one seconds then something is very wrong. So it's really hard to find

a. The right metrics to export
b. Name them correctly

because it's so very dependent on the context in which the technology is being used in.

Let's consider this:

I have a small site which isn't latency critical. But what interests me really is how often each token is being used. This metric is probably not something we'd include per default in the project.

I'm 100% sure that someone will come up with such a requirement and demand a change to the code base to add it.

Apart from the original metrics module being really shitty written, similar conceptual issues led to me removing it from the project. I thought prometheus had like some standard middleware which would be accepted as "best practice" - or just default metrics which are always collected, and use them to monitor services. But defining everything manually and (in my case) without much experience with prometheus, will lead to lots of breaking changes and constant readjustments.

Isn't there a better way to achieve this?

Reply:
By default prometheus client gathers information about:

- Number of goroutines that currently exist.
- A summary of the GC invocation durations
- Number of bytes allocated and still in use
- Total number of bytes allocated, even if freed
- Number of bytes obtained by system. Sum of all system allocations
- Total number of pointer lookups
- Total number of mallocs
- Total number of frees
- Number of heap bytes allocated and still in use
- Number of heap bytes obtained from system
- Number of heap bytes waiting to be used
- Number of heap bytes that are in use
- Total number of heap bytes released to OS
- Number of allocated objects
- Number of bytes in use by the stack allocator
- Number of bytes obtained from system for stack allocator
- Number of bytes in use by mspan structures
- Number of bytes used for mspan structures obtained from system
- Number of bytes in use by mcache structures
- Number of bytes used for mcache structures obtained from system
- Number of bytes used by the profiling bucket hash table
- Number of bytes used for garbage collection system metadata
- Number of bytes used for other system allocations
- Number of heap bytes when next garbage collection will take place
- Number of seconds since 1970 of last garbage collection
- Total user and system CPU time spent in seconds
- Number of open file descriptors
- Maximum number of open file descriptors
- Virtual memory size in bytes
- Resident memory size in bytes
- Start time of the process since unix epoch in seconds

All other metrics should be added manually. Thats why i added 2 metrics that may be needed for users

Reply:
By default prometheus client gathers information about:

- Number of goroutines that currently exist.
- A summary of the GC invocation durations
- Number of bytes allocated and still in use
- Total number of bytes allocated, even if freed
- Number of bytes obtained by system. Sum of all system allocations
- Total number of pointer lookups
- Total number of mallocs
- Total number of frees
- Number of heap bytes allocated and still in use
- Number of heap bytes obtained from system
- Number of heap bytes waiting to be used
- Number of heap bytes that are in use
- Total number of heap bytes released to OS
- Number of allocated objects
- Number of bytes in use by the stack allocator
- Number of bytes obtained from system for stack allocator
- Number of bytes in use by mspan structures
- Number of bytes used for mspan structures obtained from system
- Number of bytes in use by mcache structures
- Number of bytes used for mcache structures obtained from system
- Number of bytes used by the profiling bucket hash table
- Number of bytes used for garbage collection system metadata
- Number of bytes used for other system allocations
- Number of heap bytes when next garbage collection will take place
- Number of seconds since 1970 of last garbage collection
- Total user and system CPU time spent in seconds
- Number of open file descriptors
- Maximum number of open file descriptors
- Virtual memory size in bytes
- Resident memory size in bytes
- Start time of the process since unix epoch in seconds

All other metrics should be added manually. Thats why i added 2 metrics that may be needed for users

Reply:
Ah I see - maybe a good idea then would be to just send the default metricsf or now?

Reply:
In this case new middleware does not make sense. New endpoint will be added into "health".

Should i leave middleware as skeleton for future or it can be deleted?

Reply:
I propose to leave a empty middleware for prometheus with examples. It is given ability to simple add custom metrics if needed

Reply:
Yeah I think that makes sense - we could - in the future - also have the ability to load a plugin for handling prometheus middleware, which would make this a nobrainer to use.